autoload/lsp/ui/vim/completion.vim

```diff
@@ -0,0 +1,240 @@
+  call feedkeys(printf("\<C-r>=<SNR>%d_on_complete_done_after()\<CR>", s:SID()), 'n')
+endfunction
+
+"
+" Apply textEdit or insertText(snippet) and additionalTextEdits.
+"
+function! s:on_complete_done_after() abort
+  let l:line = s:context.line
+  let l:position = s:context.position
+  let l:completed_item = s:context.completed_item
+  let l:server_name = s:context.server_name
+  let l:completion_item = s:context.completion_item
+
+  " check the commit characters are <BS> or <C-w>.
+  if strlen(getline('.')) < strlen(l:line)
+    doautocmd User lsp_complete_done
+    return ''
+  endif
+
+  " Do nothing if text_edit is disabled.
+  if !g:lsp_text_edit_enabled
+    doautocmd User lsp_complete_done
+    return ''
+  endif
+
+  let l:completion_item = s:resolve_completion_item(l:completion_item, l:server_name)
+
+  " apply textEdit or insertText(snippet).
+  let l:expand_text = s:get_expand_text(l:completed_item, l:completion_item)
+  if strlen(l:expand_text) > 0
+    call s:clear_inserted_text(
+          \   l:line,
+          \   l:position,
+          \   l:completed_item,
+          \   l:completion_item
+          \ )
+    if exists('g:lsp_snippets_expand_snippet') && len(g:lsp_snippets_expand_snippet) > 0
+      " vim-lsp-snippets expects commit characters removed.
+      call s:expand_text_simply(v:completed_item.word)
+    elseif exists('g:lsp_snippet_expand') && len(g:lsp_snippet_expand) > 0
+      " other snippet integartion point.
+      call g:lsp_snippet_expand[0](l:expand_text)
+    else
+      " expand text simply.
+      call s:expand_text_simply(l:expand_text)
+    endif
+  endif
+
+  " apply additionalTextEdits.
+  if has_key(l:completion_item, 'additionalTextEdits')
+    let l:saved_mark = getpos("'a")
+    let l:pos = getpos('.')
+    call setpos("'a", l:pos)
+    call lsp#utils#text_edit#apply_text_edits(
+          \ lsp#utils#get_buffer_uri(bufnr('%')),
+          \ l:completion_item.additionalTextEdits
+          \ )
+    let l:pos = getpos("'a")
+    call setpos("'a", l:saved_mark)
+    call setpos('.', l:pos)
+  endif
+
+  doautocmd User lsp_complete_done
+  return ''
+endfunction
+
+"
+" Try `completionItem/resolve` if it possible.
+"
+function! s:resolve_completion_item(completion_item, server_name) abort
+  " server_name is not provided.
+  if empty(a:server_name)
+    return a:completion_item
+  endif
+
+  " check server capabilities.
+  let l:capabilities = lsp#get_server_capabilities(a:server_name)
+  if !has_key(l:capabilities, 'completionProvider')
+        \ || !has_key(l:capabilities.completionProvider, 'resolveProvider')
+    return a:completion_item
+  endif
+
+  let l:ctx = {}
+  let l:ctx.response = {}
+  function! l:ctx.callback(data) abort
+    let l:self.response = a:data.response
+  endfunction
+
+  call lsp#send_request(a:server_name, {
+        \   'method': 'completionItem/resolve',
+        \   'params': a:completion_item,
+        \   'sync': 1,
```

Comment:
any particular reason we set `sync` to `1`. If possible we should avoid using sync since this as mainly added for vim to support `LspDocumentFormatSync` since vim doesn't have autocmd for async save.

Reply:
This sync is needed by nice behavior.
I think CompleteDone should not contain async process.

If it changes to async, the user might see the below lines.

1. call getbuf|<C-x><C-o><C-n>(
2. call getbufline(
3. call getbufline(expr, lnum)

We should not show the line of case 2, I think.

FYI:
I tried `s:on_complete_done_after` call by `timer_start(0, ...)`.
It has side effect that show the line of case 2.
So I changed the implementation to use `feedkeys`.

Reply:
sounds good. This means we should make sure we have flag to disable `completionItem/resolve` for slow language servers.